File: api-server/Presentation/Controllers/ArticleController.cs
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using api_server.Contract.Requests;
using api_server.Handlers;
using api_server.Contract.Responses;
using api_server.Data;
using api_server.Presentation;

namespace api_server.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    [Authorize]
    public class ArticleController : Controller
    {
        [HttpGet]
        [AllowAnonymous]
        [Route("search")]
        public async Task<ArticlesResponse> Search(SearchArticlesRequest request, [FromServices] IHandler<SearchArticlesRequest, ArticlesResponse> handler)
        {
            return await handler.Handle(request);
        }

        [HttpGet]
        [AllowAnonymous]
        [Route("{id}")]
        public async Task<ArticleResponse> GetById(GetArticleByIdRequest request, [FromServices] IHandler<GetArticleByIdRequest, ArticleResponse> handler)
        {
            return await handler.Handle(request);
        }

        [HttpGet]
        [AllowAnonymous]
        [Route("")]
        public async Task<ArticlesResponse> GetAll(GetAllArticleRequest request, [FromServices] IHandler<GetAllArticleRequest, ArticlesResponse> handler)
        {
            return await handler.Handle(request);
        }

        [HttpPost]
        [Authorized(Role.Admin)]
        [Route("isUniqueTitle")]
        public async Task<BooleanResponse> IsUniqueTitle(IsUniqueTitleRequest request, [FromServices] IHandler<IsUniqueTitleRequest, BooleanResponse> handler)
        {
            return await handler.Handle(request);
        }

        [HttpPost]
        [Authorized(Role.Admin)]
        [Route("")]
        public async Task<IdResponse> CreateArticle(CreateArticleRequest request, [FromServices] IHandler<CreateArticleRequest, IdResponse> handler)
        {
            return await handler.Handle(request);
        }

        [HttpPut]
        [Authorized(Role.Admin)]
        [Route("{id}")]
        public async Task<IdResponse> EditArticle(EditArticleRequest request, [FromServices] IHandler<EditArticleRequest, IdResponse> handler)
        {
            return await handler.Handle(request);
        }

        [HttpDelete]
        [Authorized(Role.Admin)]
        [Route("{id}")]
        public async Task<BooleanResponse> DeleteArticle(DeleteArticleRequest request, [FromServices] IHandler<DeleteArticleRequest, BooleanResponse> handler)
        {
            return await handler.Handle(request);
        }
    }
}

File: apps/api-server/src/controllers/article.controller.ts
import {
  Param,
  Get,
  JsonController,
  Post,
  Put,
  Delete,
} from 'routing-controllers';
import executor from '../handlers/executor';
import {
  ArticleResponse,
  ArticlesResponse,
  BooleanResponse,
  IdResponse,
} from '../../../../libs/responses';
import {
  CreateArticleRequest,
  DeleteArticleRequest,
  EditArticleRequest,
  GetAllArticleRequest,
  GetArticleByIdRequest,
  IsUniqueTitleRequest,
  SearchArticleRequest,
} from '../contract/requests';

@JsonController()
export class ArticleController {
  @Get('/search')
  async search(request: SearchArticleRequest) {
    return await executor.Execute<SearchArticleRequest, ArticlesResponse>(
      nameof<SearchArticleRequest>(),
      request
    );
  }

  @Get('/:id')
  async getById(@Param('id') request: GetArticleByIdRequest) {
    return await executor.Execute<GetArticleByIdRequest, ArticleResponse>(
      nameof<GetArticleByIdRequest>(),
      request
    );
  }

  @Get('/')
  async get() {
    return 'CI worked 17!';
  }

  // @Get("/")
  // async get(request: GetAllArticleRequest) {
  //    return await executor.Execute<GetAllArticleRequest, ArticlesResponse>(nameof<GetAllArticleRequest>(), request);
  // }

  @Post('/isUniqueTitle')
  async IsUniqueTitleRequest(request: IsUniqueTitleRequest) {
    return await executor.Execute<IsUniqueTitleRequest, BooleanResponse>(
      nameof<IsUniqueTitleRequest>(),
      request
    );
  }

  @Post('/')
  async Create(request: CreateArticleRequest) {
    return await executor.Execute<CreateArticleRequest, IdResponse>(
      nameof<CreateArticleRequest>(),
      request
    );
  }

  @Put('/:id')
  async Edit(request: EditArticleRequest) {
    return await executor.Execute<EditArticleRequest, IdResponse>(
      nameof<EditArticleRequest>(),
      request
    );
  }

  @Delete('/:id')
  async Delete(request: DeleteArticleRequest) {
    return await executor.Execute<DeleteArticleRequest, BooleanResponse>(
      nameof<DeleteArticleRequest>(),
      request
    );
  }
}
